web/context.go

```diff
@@ -0,0 +1,276 @@
+/*
+Copyright 2017 Caicloud Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package web
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"mime/multipart"
+	"net"
+	"net/http"
+	"net/url"
+)
+
+var (
+	// contextKeyUnderlyingHttpContext is a key for context.
+	// It's unique and point to httpCtx.
+	contextKeyUnderlyingHttpContext interface{} = new(byte)
+)
+
+// httpCtx contains a http.Request and a http.ResponseWriter for a request.
+// It goes through the life cycle of a request.
+type httpCtx struct {
+	context.Context
+	container container
+	response  response
+}
+
+func newHttpContext() interface{} {
+	ctx := &httpCtx{}
+	ctx.container.params = make([]param, 0, 5)
+	return ctx
+}
+
+// reset resets itself by a http.Request and a http.ResponseWriter.
+func (c *httpCtx) reset(w http.ResponseWriter, request *http.Request) {
+	// Get context from request.
+	c.Context = request.Context()
+	// Reset value container.
+	c.container.request = request
+	c.container.params = c.container.params[:0]
+	c.container.query = nil
+	// Reset response.
+	c.response.writer = w
+	c.response.statusCode = 0
+	c.response.contentLength = 0
+}
+
+// clear clears request and response.
+func (c *httpCtx) clear() {
+	c.Context = nil
+	c.container.request = nil
+	c.container.query = nil
+	c.response.writer = nil
+}
+
+// Value returns itself when key is contextKeyUnderlyingHttpContext.
```

Comment:
It is not possible for a outside pkg passing `contextKeyUnderlyingHttpContext` to `Value`.

Reply:
External packages should not use httpCtx.
There are three utils for external packages:
```go
func HTTPRequest(ctx context.Context) *http.Request 
func HTTPResponseWriter(ctx context.Context) ResponseWriter 
func HTTPValueContainer(ctx context.Context) ValueContainer 
```

Reply:
Yes, but users can see it it in comments...